### ISSUE TITLE
Accessing different viewport buffers through ViewportTextures 

### ISSUE THREAD

**Issue by SIsilicon** [godotengine/godot-proposals#798]
**Describe the project you are working on:**
This proposal applies to any project involving screen space shaders.

**Describe the problem or limitation you are having in your project:**
You see, what I'd like to access the depth texture of the scene for a shader of mine. However, because shader is so expensive, I wanted to render the effect at half or even quarter resolution to composite on top of the scene. But then I would have no access to the depth texture as it would not be accessible from any other viewport.
Plus, I have another project that requires access to the material buffers (normal, metalness, etc.) for a screenspace effect, but those aren't accessible period!

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
What I propose is that user should be given access to these buffers from viewport textures, maybe like give viewport textures a `buffer_mode` (with color being the default). This way, a lot more 3D visual effects can be achieved, such as screen space global illumination, and the ability to use a viewport's depth texture from another viewport.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

Starting from this function:
https://github.com/godotengine/godot/blob/4d2c8ba922085a94ab7e963d23289fb8ed9c1f08/drivers/gles3/rasterizer_storage_gles3.cpp#L7530-L7540
we could expose the buffers to the front end. However, I also see that the buffers are set as Renderbuffer Objects. They would need to be initialized as textures in order to be accessible.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
No, 'cause this requires low level access to the rendering API.

**Is there a reason why this should be core and not an add-on in the asset library?:**
Because of the reason above.

**Comment by SIsilicon** [comment]
Also, I'm willing to do the code implementation myself.

**Comment by clayjohn** [comment]
This is planned for 4.0! Your help would be very welcome. Please join the developer IRC channel so you can discuss the best way to implement it with Juan.

**Comment by SIsilicon** [comment]
Well, I'm actually talking about 3.2.
Don't get me wrong. I would love to help out with Godot 4.0's rendering API, but I don't have much experience with Vulkan, and even if I did, I don't have the hardware to even test it yet.
The feature I was talking about, I was speaking for Godot 3.2. I know 4.0 will come along with more features, but I wanted to see if it's possible to add some to 3.2, while it's still at least relevant.

**Comment by SIsilicon** [comment]
Plus I wouldn't worry about inconsistencies between versions. After all, there will be _lots_ of changes and features in 4.0. Perhaps it might even be possible to port this feature up to 4.0, if the pull request goes through.

**Comment by clayjohn** [comment]
In general we won't add new features to previous releases. The reason for that is that new features typically introduce a lot of usability problems and bugs and the idea behind the maintenance releases is that things get more stable instead of less stable. There are exceptions. For example, batching is being added to the GLES2 2D renderer. Batching is a much needed feature that just wasn't ready for the 3.2 release and in order to merge it, we are painstakingly testing it in beta builds.

Another catch is that the 3.2 rendering API wasn't created with this proposal in mind and is not very flexible. So adding access to the various buffers will be very difficult to implement in an elegant way and will be especially prone to creating bugs and breaking user's projects.

That being said, since you are willing to do the work yourself you could always fork the engine and do the work. If you are able to do it in a clean way that doesn't look too error prone we could consider merging it in (but this would be unlikely). If it doesn't get merged in, you would have to maintain your own fork (which shouldn't be too hard because we won't be making substantial changes to the 3.2 branch).

**Comment by SIsilicon** [comment]
Noted. I'm sure I can figure out a way of implementing this without introducing bugs.

**Comment by clayjohn** [comment]
I would love to see it :) ping me when you have something and I will be glad to discuss it with you.

**Comment by SIsilicon** [comment]
Hey hey! I have some progress to show today. The code works great, and the buffers are now accessible.
![ezgif com-video-to-gif (1)](https://user-images.githubusercontent.com/34734122/82155794-0efb5200-983d-11ea-9609-896687680d94.gif)
The depth buffer works fine by the way. It just looks all red because of how its data is stored. Speaking of depth buffer, it's data is currently inaccessible through `get_data` for now, since its a 24-bit depth format, but I think that's OK for now. I also noticed that some of the buffers are only accessible when post process effects like SSR and SSAO are enabled, so I also added a `force_mrt` variable to `Viewport` incase you want to access those buffers without enabling those heavy effects, but I haven't tested it yet. Just a few more days of bug ironing. :)

**Comment by clayjohn** [comment]
That looks great! Indeed, mrts have to be enabled for the various buffers to get used. Using MRTs will also add some overhead.

**Comment by golddotasksquestions** [comment]
Would be nice if depth would be displayed as grey scale. To be honest I don't understand how it shows up as solid read yet is supposed to store the depth buffer. I only know depth buffers are greyscale image. If the red channel is supposed to be used as the depth buffer, we would not see a solid color, be would see red fading to transparent.

In the studio we often just screenshot or print screen and use that for iterating on concept art and level design. If it would show up as greyscale it would make iteration flow faster and would also make more visual sense to artists creating visual effects.

**Comment by TwistedTwigleg** [comment]
> Would be nice if depth would be displayed as grey scale. To be honest I don't understand how it shows up as solid read yet is supposed to store the depth buffer. I only know depth buffers are greyscale image. If the red channel is supposed to be used as the depth buffer, we would not see a solid color, we would see red fading to transparent.

You can display the depth buffer as grey scale by taking the red channel and coping its values to both the green and blue channels, making the depth display monochrome. That said, from a purely performance perspective, leaving it red might be better since it avoids having to set two additional properties per pixel.

**Comment by SIsilicon** [comment]
> Would be nice if depth would be displayed as grey scale. To be honest I don't understand how it shows up as solid read yet is supposed to store the depth buffer. I only know depth buffers are greyscale image. If the red channel is supposed to be used as the depth buffer, we would not see a solid color, we would see red fading to transparent.
> 
> In the studio we often just screenshot or print screen and use that for iterating on concept art and level design. If it would show up as greyscale it would make iteration flow faster and would also make more visual sense to artists creating visual effects.

Here's how the depth buffer works. It only stores it's data in one channel. The fact that it's red is irrelevant. You can easily display it as greyscale by doing this.

```
COLOR.rgb = texture(tex, uv).rrr;
```

Interestingly though it _does_ show up as greyscale when working in GLES2.

Also, the reason why it _appears_ to be solid red is because of the camera's mode being `Perspective`. Most of the information is packed close to the camera's near plane, and you'd have to do some path to make it linear. TLDR, the depth buffer should work exactly the same as the built-in `DEPTH_TEXTURE` variable in spatial shaders.

**Comment by golddotasksquestions** [comment]
> The fact that it's red is irrelevant. You can easily display it as greyscale by doing this.

No, it's not irrelevant. 
You can't screenshot the red channel and use it immediately.
You also can't assume artists know or understand shader code. This may be easy to implement in a shader flow, but for multidisciplinary production flow, this is incredibly user unfriendly.

**Comment by SIsilicon** [comment]
Ok, I think I could get it to look white out of the box. I could some "texture swizzling" to look white. However, the packed nature of the buffer must stay.

**Comment by TwistedTwigleg** [comment]
@SIsilicon 
I'm not sure how hard it would be, but maybe make an optional depth mode that converts the texture from red to white? That way, those who need the grayscale image can get it, but those who do not can just use the raw depth buffer. That said, I'm not sure how much code complexity it would add, so maybe it is better to just have a single depth mode.

_________

@golddotasksquestions

You can convert a completely red depth image to grayscale in a tool like Photoshop, and then use that as the depth. It is just an added step, but most tools that have basic color manipulation should allow for the needed color manipulation.

For example, here is the process I used on [this image](http://thedemonthrone.ca/projects/rendering-terrain/rendering-terrain-part-13-basic-shadow-maps-the-not-quite-right-edition/):

![ShadowMapDepthTest_Grayscale_Example](https://user-images.githubusercontent.com/25082678/82249049-23ac1880-9917-11ea-946a-4cd665852590.jpg)

* Import the red image
* Click the Colors drop down, and select "Hue-Saturation"
* In the window that opens, set the "Saturation" slider to -100
* This will make the image grayscale 🙂

Granted, this is an additional step, However, it may be visual art software, like those you mentioned, use the red channel of the image, as it is commonly the first channel in a texture and depth only needs to be stored in one channel.
Based on [this page](https://helpx.adobe.com/photoshop/using/adjusting-image-sharpness-blur.html) from Adobe's help about image sharpness and [this page](https://helpx.adobe.com/photoshop/using/channel-basics.html) about Photoshop channel basics, it also looks like you can selectively only make a single channel visible, and doing this for the red channel would result in an automatically grayscale image, if I understand the page correctly.

**Comment by SIsilicon** [comment]
@TwistedTwigleg 
Don't worry, it shouldn't be hard at all. Just needed to add these two lines.
```
GLint swizzleMask[] = { GL_RED, GL_RED, GL_RED, GL_ONE };
glTexParameteriv(GL_TEXTURE_2D, GL_TEXTURE_SWIZZLE_RGBA, swizzleMask);
```
Plus, adding this makes the depth buffer look similar to the one in GLES2, which I previously stated was also grey-scale.

**Comment by TwistedTwigleg** [comment]
Awesome! 👍

**Comment by golddotasksquestions** [comment]
@TwistedTwigleg Thanks, I appreciate the effort you put into explaining the steps in detail, but I'm afraid you are explaining this to the wrong person, I'm using Photoshop since 23 years.

**Comment by SIsilicon** [comment]
Well I finally created a [pull request](https://github.com/godotengine/godot/pull/38926), but some of tests have failed.

**Comment by TwistedTwigleg** [comment]
Awesome!

________

From what I can deduce from the [Travis error](https://travis-ci.org/github/godotengine/godot/jobs/689695760), it seems there is code that is not formatted with the correct clang format, and that is causing the issue.

**Comment by SIsilicon** [comment]
Well I'm confused. Can someone explain to me what this is supposed to mean?
![](https://user-images.githubusercontent.com/34734122/82582013-06658d00-9b57-11ea-87c9-186be1099e06.png)

**Comment by TwistedTwigleg** [comment]
I would say what it means, but honestly I have no idea! I knew at one point what it meant, but its been long enough that I forgot. I know the `+` represents additional changes/lines and the `-` means removed lines, but I am not sure which needs to be used in this case to make it compliant with the clang format.
I *think* there is a way to have clang-format run and make the adjustments automatically as a pre-commit hook.

**Comment by Megalomaniak** [comment]
I think what @golddotasksquestions really wants is a `Display Depth` option added to the viewport display menu:

<details>
  <summary>image of the menu</summary>
  
  ![Viewport-display-menu](https://user-images.githubusercontent.com/3162126/82656449-75f97c00-9c2c-11ea-8cb0-bbe3dd146413.png)
  
</details>

Buffer itself can continue to be in the red channel just fine, only the display preview needs to be in gray-scale. I do feel like it is a separate feature from this though.

**Comment by golddotasksquestions** [comment]
@Megalomaniak 
I don't think it's a separate feature. I think it's a question whether you want to "Display the values" (greyscale) or "Display the result" (red channel as red, green channel as green, blue channel as blue, a mix between those as an additive mix).
Since I'm coming from visual development, I'm used to "Display the values", since I this allows to directly use that greyscale image elsewhere, be it as mask or channel.

**Comment by SIsilicon** [comment]
Concerning this proposal, that _would_ be a separate feature. Such an option would require its own debug drawing mode, just like the other debug modes (unshaded, overdraw, wireframe).

**Comment by PranavSK** [comment]
So wondering if this will get picked for the 3.2 updates. Also, how would the buffers be accessible from the shaders? Right now it's possible to get a `SCREEN_TEXTURE` and `DEPTH_TEXTURE` inside the shader. Maybe the `SCREEN_TEXTURE` is dependent on the mode of the viewport?

**Comment by Megalomaniak** [comment]
> 
> 
> @Megalomaniak
> I don't think it's a separate feature. I think it's a question whether you want to "Display the values" (greyscale) or "Display the result" (red channel as red, green channel as green, blue channel as blue, a mix between those as an additive mix).
> Since I'm coming from visual development, I'm used to "Display the values", since I this allows to directly use that greyscale image elsewhere, be it as mask or channel.

No I mean the whole display (of whichever) is a separate feature that would or at least could follow this one. But I do agree that it would be useful of course. I'd recommend opening another proposal for it.

**Comment by SIsilicon** [comment]
> So wondering if this will get picked for the 3.2 updates. Also, how would the buffers be accessible from the shaders? Right now it's possible to get a `SCREEN_TEXTURE` and `DEPTH_TEXTURE` inside the shader. Maybe the `SCREEN_TEXTURE` is dependent on the mode of the viewport?

You make a good point. My PR currently doesn't expose those textures in shaders. Probably be a good idea to do so, but I'd rather leave that for if/when it gets merged.

**Comment by ceLoFaN** [comment]
Is there any plan for a similar functionality to be implemented in 4.0?

**Comment by Calinou** [comment]
> Is there any plan for a similar functionality to be implemented in 4.0?

This may be implemented only after 4.0 depending on time constraints.

**Comment by Zylann** [comment]
I have a shader here which "magically" adds a small bevel to edges, which is very useful when creating hard-surface objects out of blocks, without having to deal with neighbor management or geometry:

![image](https://user-images.githubusercontent.com/1311555/140620963-e4129e68-c13f-410b-b6a7-f4519cccda72.png)

This would need access to a normals buffer, which may contain, at minimum, the surfaces having that material (or all of them would work too if such buffer is already available).

In a nutshell, this picks 12 samples from the normals buffer and averages them. There are several edge cases to adress though for it to be performant enough and free of unwanted artifacts.

**Comment by nsrosenqvist** [comment]
(off topic) @Zylann would love to have access to that shader source if you're willing to share it?

**Comment by Calinou** [comment]
Note that 4.0 exposes `NORMAL_ROUGHNESS_TEXTURE`, so you can access both view-space normals and the roughness buffer.

**Comment by Megalomaniak** [comment]
> @SIsilicon [said](https://github.com/godotengine/godot-proposals/issues/798#issuecomment-632219024): "Well I'm confused. Can someone explain to me what this is supposed to mean?"

Sorry, 2 years late, missed it back then. That just looks like a line-break/space was removed and another added. Nothing to be worried about.

**Comment by MackyK** [comment]
Any chance of getting this for 4.0?

It's required for project.

**Comment by Calinou** [comment]
> Any chance of getting this for 4.0?
> 
> It's required for project.

Which buffers do you need access to? Normal and roughness are exposed as I mentioned in https://github.com/godotengine/godot-proposals/issues/798#issuecomment-1211802282.

**Comment by MackyK** [comment]
> > Any chance of getting this for 4.0?
> > It's required for project.
> 
> Which buffers do you need access to? Normal and roughness are exposed in 4.0.beta as I mentioned in [#798 (comment)](https://github.com/godotengine/godot-proposals/issues/798#issuecomment-1211802282).

I need Depth Buffer and Stencil Buffer.

**Comment by michael-nischt** [comment]
Having the same/similar issue here. Being blocked on Godot 4.0 beta.

Basically we need the depth buffer of a view-port for a custom post-fx _together_ with the _final color_ image (of the view-port only, without UI ..). Ideally with nice way to have access to the projection / view matrices of the view-port / camera but I guess that's possible to construct manually.

The important part for us is that it works with transparent objects and the the depth image contains their values (or not) based on the material setting:

![image](https://user-images.githubusercontent.com/500820/201932850-a7f59877-191d-4d5a-9393-f8e2840f6abf.png)

**Comment by Calinou** [comment]
> The important part for us is that it works with transparent objects and the the depth image contains their values (or not) based on the material setting:

You need to use the opaque prepass mode for this, and this will only work on fully opaque parts of the material:

https://github.com/godotengine/godot/issues/68356

There's no way *translucent* materials can be present in the depth buffer.

**Comment by prominentdetail** [comment]
What is needed is a way to access the depth data from the transparent pass, right? Is there a way to access that or store it so that it can be read in a shader?

**Comment by Zireael07** [comment]
@prominentdetail No, as depth data is captured earlier, before transparent pass. This limitation is documented after some people ran into it, including yours truly

Workaround: use a material dither shader, there's quite a few of them floating around the net

**Comment by michael-nischt** [comment]
To solve our current problem it is ok(ish) to have only the opaque (pre-pass) depth as long as we have transparent objects in the color (screen) image.

Is it already possible to access the pre-pass depth in a custom post-fx (CanvasItem) shader?
I can't find a [built-in](https://docs.godotengine.org/en/latest/tutorials/shaders/shader_reference/canvas_item_shader.html) variable.

----

Out of curiosity: How does does the Environment's Screen-Space Reflection work with _translucent_ objects which are highly secular? 
Does it use the depth (and normal) from the opaque object behind (incorrect)?
Or does the SSR run before the transparent pass (just not reflecting of transparent objects)?
Or is there an threshold to define _almost_ opaque fragments for which alpha values above are written to the depth buffer  (e,g. using [glFragDepth](https://registry.khronos.org/OpenGL-Refpages/gl4/html/gl_FragDepth.xhtml)?

**Comment by Megalomaniak** [comment]
> There's no way _translucent_ materials can be present in the depth buffer.

Well, via dithering and stochastic filtering I think inferred(a type of deferred) renderers have done this before.

**Comment by sserafimescu** [comment]
Accessing a viewport's the depth buffer while rendering to another viewport is essential for many types of effects. For example, rendering volumetric effects at a lower resolution while using the main scene's depth. I am currently being blocked by this missing feature, and I really hope it will make it to a 4.x release soon.

**Comment by CaveJohnson376** [comment]
gosh, this feature is a live saver for me. i quite literally dreamed about it for a while now!

that being said, i have several projects on hold, where i have interiors and exteriors rendered as separate passes, and i combine them into one frame, but without access to depth textures of both it's quite hard to correctly combine them.

oh, and it could also allow to emulate deferred lighting/shading, but only in theory - info like shadow maps, light positions and matrices are not exposed in gdshader, only fragment's direction towards light, but that all is completely different story...

**Comment by ttencate** [comment]
I also have a use case for this, that I don't think is covered by existing functionality in 4.1.1.

I want to implement a kind of fog-of-war in a 3D scene, which will make everything black that is farther away than some distance _d_ from some point _p_ in world space (which does _not_ coincide with the camera position).

Currently, I'm using a screen-reading shader for this, which applied to a full-screen quad. Something along the lines of:

```
shader_type spatial;
render_mode unshaded, depth_draw_never, depth_test_disabled;

uniform sampler2D screen_texture : hint_screen_texture, filter_nearest;
uniform sampler2D depth_texture : hint_depth_texture, filter_nearest;

uniform vec3 view_position;
uniform float view_distance;
uniform vec3 fog_color : source_color;

void fragment() {
	vec3 color = textureLod(screen_texture, SCREEN_UV, 0.0).rgb;
	float depth = textureLod(depth_texture, SCREEN_UV, 0.0).r;

	vec4 fragment_position_h = INV_VIEW_MATRIX * INV_PROJECTION_MATRIX * vec4(SCREEN_UV * 2.0 - 1.0, depth, 1.0);
	vec3 fragment_position = fragment_position_h.xyz / fragment_position_h.w;
	
	float fragment_distance = distance(view_position, global_position);
	float fog_density = smoothstep(0.9 * view_distance, view_distance, fragment_distance);
	ALBEDO = mix(color, fog_color, fog_density);
}
```

The problem with this is that transparent objects are rendered _after_ the `screen_texture` and `depth_texture` are captured, so they don't obey the fog-of-war.

Rendering the entire scene (including transparent objects) to a viewport, then applying a shader like this to render the viewport's contents to the screen, would solve that nicely.

(Transparent objects would be fogged according to the distance to whatever is behind them. I think that's fine for my use case, but if it becomes a problem, I could just force transparent objects to write to the depth buffer as well.)

**Comment by nightrune** [comment]
Adding my somewhat different use case. I'm wanting to do raycasting with voxels. I'm looking to be able to reuse the depth buffer with velocity between scenes to save casting occluded objects. It would also let me play with raycast in smaller resolution and upscaling

**Comment by OhiraKyou** [comment]
My use case for this is downward projection of a character's shadow. You capture separate depth buffers for the environment and the player, from above. Then, you get the world position of the main camera's depth. With all of that data, a cube decal can render a projected character shadow exclusively on the top-most opaque environment surface.

**Comment by samanthabeilman** [comment]
Any chance that something like this will be coming in 4.x?

**Comment by Calinou** [comment]
> I am now wondering when we might hope to see it! <3

4.3 is currently in feature freeze, so this feature can only make it to 4.4 *at the earliest*. There is a partial implementation of this proposal in https://github.com/godotengine/godot/pull/89196 which you could [test](https://docs.godotengine.org/en/latest/contributing/workflow/testing_pull_requests.html) right now, but it doesn't address all use cases (such as accessing the normal/roughness buffer).

**Comment by ffdd270** [comment]
Hello. I was wondering if there have been any updates on this issue.

I’m currently working on a shader that renders specific objects separately and then composites them back into the scene, but I haven’t been able to find a direct way to access the depth buffer.

Since this issue seems closely related, I wanted to ask about it.

It appears to be a fairly essential feature, so I was a bit surprised that there haven’t been updates for about two years.

(I’m not trying to rush anything—just wondering if there might be alternative approaches to this problem, and whether that’s why it hasn’t been prioritized.)

**Comment by ttencate** [comment]
@ffdd270 Accessing the _depth_ buffer from a shader is totally possible. Look up the section on screen-reading shaders in the docs. Other buffers apart from the color buffer are still inaccessible.

There's also the new experimental compositor effects, that require some more legwork with compute shaders but have a lot of power.

**Comment by ffdd270** [comment]
@ttencate 

I realize now that what I actually wanted wasn’t the screen’s depth, but the depth values from SubViewport A and SubViewport B—separate from the color values rendered by each camera. So I should have simply added A’s depth and B’s depth as additional inputs and handled them accordingly. It turned out to be a rather naive question. Thank you for your answer. Have a nice day.

**Comment by OhiraKyou** [comment]
> [@ttencate](https://github.com/ttencate)
> 
> So I should have simply added A’s depth and B’s depth as additional inputs and handled them accordingly.

As the topic describes, however, the depth of any given viewport is only accessible when rendering to that viewport; it is inaccessible from others. So, that excludes using two SubViewport depths to calculate anything else. For example, [the use case I previously described](https://github.com/godotengine/godot-proposals/issues/798#issuecomment-1744017134) would also use two SubViewport depth buffers (to calculate the color of shadow projector decal volume).
